sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -189,6 +189,9 @@ abstract class Optimizer(catalogManager: CatalogManager)
     // plan may contain nodes that do not report stats. Anything that uses stats must run after
     // this batch.
     Batch("Early Filter and Projection Push-Down", Once, earlyScanPushDownRules: _*) :+
+    // This batch rewrites plans for v2 tables. It should be run after the operator
+    // optimization batch and before any batches that depend on stats.
+    Batch("V2 Source Rewrite Rules", Once, v2SourceRewriteRules: _*) :+
```

Comment:
If these rules create scan nodes, then shouldn't this run before the early filter and projection pushdown? Then we can rewrite to normal logical plans and have those rules perform early pushdown like normal.

Reply:
Right now, scan construction and pushdown happens in the same method. We could split that later so I agree with moving this batch before early filter and projection pushdown.

Reply:
Done.